transformations/transformations/parallel_routine_dispatch.py

```diff
@@ -56,3 +99,146 @@ def create_dr_hook_calls(scope, cdname, pkey):
+            handle=sym.Variable(name='ZHOOK_HANDLE_FIELD_API', scope=routine)
+        )
+        routine.body.insert(2, (dr_hook_calls[0], ir.Comment(text=''), *self.new_calls, dr_hook_calls[1]))
+
+        # Insert the field deletion wrapped into a DR_HOOK call
+        dr_hook_calls = self.create_dr_hook_calls(
+            routine, cdname='DELETE_TEMPORARIES',
+            handle=sym.Variable(name='ZHOOK_HANDLE_FIELD_API', scope=routine)
+        )
+        routine.body.insert(-2,(dr_hook_calls[0], ir.Comment(text=''), *self.delete_calls, dr_hook_calls[1]))
+
+    def decl_derived_types(self, routine, region):
+        region_map_derived = {}
+        derived = [var for var in FindVariables().visit(region) if var.name_parts[0] in routine.arguments]
+        for var in derived :
+            
+            key = f"{routine.variable_map[var.name_parts[0]].type.dtype.name}%{'%'.join(var.name_parts[1:])}"
+            if key in self.map_index:
+                value = self.map_index[key]
+                # Creating the pointer on the data : YL_A
+                data_name = f"Z_{var.name.replace('%', '_')}"
+                if "REAL" and "JPRB" in value[0]:
+                    data_type = SymbolAttributes(
+                        dtype=BasicType.REAL, kind=routine.symbol_map['JPRB'],
+                        pointer=True
+                    )
+                    data_dim = value[2] + 1
+                    data_shape = (sym.RangeIndex((None, None)),) * data_dim
+                    ptr_var = sym.Variable(name=data_name, type=data_type, dimensions=data_shape, scope=routine)
+
+                else:
+                    raise NotImplementedError("This type isn't implemented yet")
+
+                # Creating the pointer on the field api object : YL%FA, YL%F_A...
+                if routine.variable_map[var.name_parts[0]].type.dtype.name=="MF_PHYS_SURF_TYPE":
+                    # YL%PA becomes YL%F_A
+                    field_name = f"{'%'.join(var.name_parts[:-1])}%F_{var.name_parts[-1][1:]}"
+                elif routine.variable_map[var.name_parts[0]].type.dtype.name=="FIELD_VARIABLES":
+                    # YL%A becomes YL%FA
+                    field_name = f"{'%'.join(var.name_parts[:-1])}%F{var.name_parts[-1]}"
+                    if var.name_parts[-1]=="P": #YL%FP = YL%FT0
+                        field_name = f"{field_name[-1]}T0"
+                else:
+                    # YL%A becomes YL%F_A
+                    field_name = f"{'%'.join(var.name_parts[:-1])}%F_{var.name_parts[-1]}"
+                field_ptr_var = var.clone(name=field_name)
+                region_map_derived[var.name] = [field_ptr_var, ptr_var]
+        return(region_map_derived)
+
+    def add_derived(self, routine):
+        ptr_var=()
+        for value in self.routine_map_derived.values():
+            dcl = ir.VariableDeclaration(
+                symbols=(value[1],)
+            )
+            ptr_var += (dcl,)
+        routine.spec.append(ptr_var)
```

Comment:
I would recommend using the following pattern to add the relevant declarations:
`routine.variables += tuple(v[1] for v in self.routine_map_derived.values())`